src/v/storage/readers_cache.cc

```diff
@@ -339,11 +345,33 @@ void readers_cache::dispose_in_background(entry* e) {
 
 ss::future<> readers_cache::maybe_evict() {
     auto now = ss::lowres_clock::now();
+    /**
+     * First try to evict invalid and outdated readers
+     */
     co_await evict_if([this, now](entry& e) {
         const auto invalid = !e.reader->is_reusable() || !e.valid;
         const auto outdated = e.last_used + _eviction_timeout < now;
         return invalid || outdated;
     });
+
+    // cap the leaders size
+    evict_size();
```

Comment:
I'm wondering if this would've prevented the issues we saw in the cluster earlier. IIUC, the issues happened because the time-based eviction wasn't happening fast enough to keep memory in check. In that scenario, were the readers actually being reused frequently? If not, does that mean time-based eviction and size-based eviction would be equally ineffective, given they happen based on the eviction timeout?

Reply:
i increased the frequency of timer firing but kept the timeout intact, so in this case the size based eviction will happen before the reader will be evicted based on the timeout.

Reply:
Ah :facepalm: got it, I mistook that for having a 30 +- ~1s (30/20).

I am a little wary of the timer going off more frequently though, especially give it is running on every partition.

Perhaps we can only use the aggressive timer if there are more than 200 readers? Or use condition variable and make the eviction timer be a loop instead that waits for 30 seconds, but also gets triggered when there are more than 200 readers are in the cache?

Reply:
I went over that once again and it seems that in most of the cases the evict operation will simply be a no-op. I will add background eviction call to be a part of `put()` method.

Reply:
This way we won't have to make timer firing more often.